cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -847,7 +847,8 @@ ParquetDatasetFactory::CollectParquetFragments(const Partitioning& partitioning)
     auto row_groups = Iota(metadata_subset->num_row_groups());
 
     auto partition_expression =
-        partitioning.Parse(StripPrefixAndFilename(path, options_.partition_base_dir))
+        partitioning
+            .Parse(StripPrefixAndFilename(path, options_.partition_base_dir).directory)
```

Comment:
(Also, can we cover this with a test?)

Reply:
Now that we are using the `PartitionPathFormat` as the argument, the `Parse()` method will work accordingly as per the partitioning mode. Any particular test you want here?

Reply:
Making sure that a filename partitioning works properly in this path, basically, since before it seems like it would have failed since the filename was being omitted.

Reply:
I have added a round-trip test in PyArrow to check whether the partitions are read correctly. Do we need any tests other than that?